multiqc/modules/ngsderive/ngsderive.py

```diff
@@ -455,3 +447,80 @@ def add_junctions_data(self):
+                "endedness": endedness_data.get("Endedness"),
+            }
+            if "Reads per template" in endedness_data:
+                rpt_present = True
+
+        general_headers = OrderedDict()
+        general_headers["endedness"] = {
+            "title": "Predicted Endedness",
+            "description": "Predicted library endedness from ngsderive",
+        }
+        self.general_stats_addcols(general_data, general_headers)
+
+        headers = OrderedDict()
+        headers["FyLn"] = {
+            "title": "f+l- count",
+            "description": "Number of reads with the 'first in template' FLAG set and the 'last in template' FLAG unset",
+            "format": "{:,d}",
+        }
+        headers["FnLy"] = {
+            "title": "f-l+ count",
+            "description": "Number of reads with the 'first in template' FLAG unset and the 'last in template' FLAG set",
+            "format": "{:,d}",
+        }
+        headers["FnLn"] = {
+            "title": "f-l- count",
+            "description": "Number of reads with the 'first in template' FLAG unset and the 'last in template' FLAG unset",
+            "format": "{:,d}",
+        }
+        headers["FyLy"] = {
+            "title": "f+l+ count",
+            "description": "Number of reads with the 'first in template' FLAG set and the 'last in template' FLAG set",
+            "format": "{:,d}",
+        }
+        if rpt_present:
+            headers["RPT"] = {
+                "title": "Reads per template",
+                "description": "Average number of reads per QNAME template",
+                "format": "{:,.4f}",
+            }
+        headers["endedness"] = {
+            "title": "Predicted Endedness",
+            "description": "Predicted library endedness from ngsderive",
+        }
+
+        table_data = {}
+        for sample, endedness_data in self.endedness.items():
+            table_data[sample] = {}
+            table_data[sample]["FyLn"] = int(endedness_data.get("f+l-"))
+            table_data[sample]["FnLy"] = int(endedness_data.get("f-l+"))
+            table_data[sample]["FnLn"] = int(endedness_data.get("f-l-"))
+            table_data[sample]["FyLy"] = int(endedness_data.get("f+l+"))
+            table_data[sample]["RPT"] = float(endedness_data.get("Reads per template"))
+            table_data[sample]["endedness"] = endedness_data.get("Endedness")
+
+        # Config for the plot
+        config = {
+            "id": "ngsderive_endedness_plot",
+            "title": "ngsderive: Endedness",
+            "namespace": "ngsderive",
+        }
+
+        self.add_section(
+            name="Endedness",
+            anchor="ngsderive-endedness",
+            description="""Predicted library endedness provided by ngsderive. For more information, please see
+            [the documentation](https://stjudecloud.github.io/ngsderive/subcommands/endedness/).""",
+            plot=table.plot(table_data, headers, config),
```

Comment:
Is it possible to represent these counts as a stacked bar plot, instead of a table? They're all subsets of a whole, right? Or could they sum to > 100%?

Reply:
They do sum to 100%, so a stacked bar plot is possible. I'm just afraid it might be less informative than the raw numbers. Which is admittedly counter-intuitive. Let me explain how these data will be used.

These data are more or less meant as sanity checks. The patterns you expect to see are:
Paired-End sample: `f+l-` == `f-l+` && `f+l+`==`0` && `f-l-`==`0`
Single-End sample: `f+l+` > `0` && `everything else == 0`
So in my mind, the easiest way to check those assumptions is with a table, where I can quickly glance and say "Great, these numbers are the same number" or "O no, something is greater than zero". The degree by which a category is off from it's assumed value doesn't really matter that match.

In practice, I'm running this on our data and seeing some off-by-one cases. I'm afraid that presenting this as a barplot would mask an off-by-one case, because in a plot the difference between 3million and 3million+1 would be impossible to see. But in a table, super easy to see a value is off by one.

I could see using the conditional formatting available to tables to color-code some of these cases. Would that be possible? First instinct is "Give each unique numerical value a distinct color" (there will be at most 4). Equal cells will have the same color. Unequal cells would have some contrast. That should facilitate the "at-a-glance" assumption checking.

So far I've ignored the `Reads Per Template` value. That is supposed to be exactly `2.0` (Paired-End) or exactly `1.0`(Single-End). Deviation from those exact values is what's concerning. Any ideas for how to present that more intuitively?

Reply:
I agree that we should keep the raw numbers instead of rounding or plotting them. But perhaps, indeed, these checks can be automated, so conditional formatting to highlight samples that break the condition is a good direction to look into.

Another approach would be to just add one more derived metric into the general stats that would simply show if the sample passes those conditions or not.

Reply:
It would be also helpful to add a "bad" example into the test data, where a sample doesn't pass the endedness criterias.

Reply:
Let me complicate things for us: There is a parameter to this command which enables only processing a subset of the BAM file. When this mode is active, there is obviously some amount of deviation from the _exact_ conditions I've outlined previously, just due to the randomness inherent in only processing a portion of the available reads. When run in this mode, the user supplies a tolerance range for determining the conditions. e.g. we can still call the sample `Paired-End` so long as `0.45 < (f+l-)/((f+l-)+(f-l+)) < 0.55`. This is in contrast to the typical case, when every read is processed, that requires `(f+l-)==(f-l+)`. Similarly RPT gets rounded to the nearest `int` during checks when run in "subset mode".

Generally we would discourage users from using this mode as it will mask errors. But we felt the need to enable it for consistency with the rest of the `ngsderive` commands. As you can imagine, checking out every read is the file takes a long time. A user has the option to get a decent enough guess in a few seconds if that's what their use-case calls for.

> Another approach would be to just add one more derived metric into the general stats, that would simply show if the sample passes those conditions or not. Similarly formatted to the existing "predicted instrument" and "predicted endedness" columns.

Because of the inconsistency outlined above, MultiQC can't get into the business of checking these conditions itself. I would not want to imply "conditions met" or "conditions unmet" with colors such as "green" and "red". The conditions checking is already done and presented by the value of `Endedness`. That will be `Unknown` if no conditions are met (or `Paired-End` or `Single-End` were _those_ conditions met). I wouldn't be opposed to coloring `Unknown` red, but I'm not sure it really adds a whole lot. Same for coloring `Paired-End` and `Single-End` green. I could take it or leave it.

I think my first instinct:
>  First instinct is "Give each unique numerical value a distinct color" (there will be at most 4). Equal cells will have the same color. Unequal cells would have some contrast. That should facilitate the "at-a-glance" assumption checking.

is still our best bet. I haven't come up with anything better. It wouldn't be particularly useful for the niche case that only a subset of the BAM was processed, but I think that's fine. We can't make everyone happy, so let's focus on the predominant use-case.

I wouldn't want to color the RPT cell because it is allowed to deviate some. I suppose `>=3` would be universally "a problem" today, but in the future we might find a valid use case for that (circular templates?). So to future-proof, I wouldn't want to call any RPT value "bad".

Reply:
> It would be also helpful to add a "bad" example into the test data, where a sample doesn't pass the endedness criteria.

https://github.com/ewels/MultiQC_TestData/blob/master/data/modules/ngsderive/test.endedness.without_rpt.txt

This represents the common "off by one" error case

Reply:
Wow, this is a fairly atypicaly module 😆  It would be great to write all of this in the description / help text above the table, so that users can read how about to interpret the results.

I can't think of any other modules doing this, but it should be possible to get what you want by setting `scale` to `False` and `bgcols` to a dynamically generated dict of discovered read counts + assigned colours. Then you can effectively colour code the table cells according to their numerical status.

If the actual number isn't very interesting, then I think this makes sense for me. As the normal cell background bars don't really tell you anything.